### ISSUE TITLE
Device Timestamp

### ISSUE THREAD

**Issue by MLopezJ** [MLopezJ/LwM2M-Asset-Tracker#3]
# How to report device timestamps? 

## TL;DR
The timestamps are going to be reported by the related resource of the LwM2M object. In case the object has not time related resource, the reported timestamp by the server is going to be used. 

Use content format as an way to send timestamp data is not an option because Asset Tracker v2 does not implement it in that way. Asset Tracker v2 uses resource from LwM2M objects to report timestamps. https://github.com/MLopezJ/LwM2M-Asset-Tracker/issues/2

## More

Asset Tracker web app is expected to receive an input with a defined data format, which is described [here](https://github.com/NordicSemiconductor/asset-tracker-cloud-docs/blob/saga/docs/cloud-protocol/state.reported.azure.json). It is needed that those timestamp values come from the device and not from the server https://github.com/MLopezJ/LwM2M-Asset-Tracker/issues/2 .

From the device perspective I can think in 2 options to accomplish that:
 1. Uses the stablished resource from the object specification 
 2. Uses the content format from request

### Uses the stablished resource from the object specification  
Some objects as Temperature (3303), Humidity (3304), Pressure (3323) and Push Button (3347), have a resource (`5518`) for inform timestamp data. There are 2 issues with this alternative:
* not all objects have this resource
* Coiote is not reporting 5518 resource to Device Twin in azure

### Uses the content format from request
LwM2M accepts multiple formats as part of the content-format options on its request. LwM2M+Json is one of those and it has defined a format that allows to send time values on each request. However, that data is been ignored by Coiote.  

### Problem
~**so even with those 2 options is not possible to send timestamp from the device because for example, the battery timestamp, which is required in Asset Tracker web app, is taken from the Device object and it has not the resource `5518` on it specification; also is not possible to send the timestamp using the content format specifications in the request and/or response.**~

Seems like Coiote is not supporting LwM2M v1.1 and there are some object that only have timestamps properties on the 1.1 version of the definition. https://github.com/MLopezJ/LwM2M-Asset-Tracker/issues/4

**Comment by MLopezJ** [comment]
Timestamps required by Asset Tracker web app

- Battery 
- Enviroment (**)
- GNSS
- Device (**)
- Roam (**)

> (**) = has the same value. It could mean that the value is comming from the same variable in client [here](https://github.com/NordicSemiconductor/asset-tracker-cloud-docs/blob/saga/docs/cloud-protocol/state.reported.azure.json#L12C11-L12C24)

In  this [data transition](https://github.com/MLopezJ/nRF-Asset-Tracker-through-Coiote-flow#data-transicion) is possible to see that:
* Battery value is taken from Device Object (3)
* Env values are taken from (3303), (3304) and (3323)
* GNSS values are taken from Location Object (6)
* Device value are taken from Device Object (3)
* Roam valueas are taken from Connectivity Monitoring Object (4)

Is important to notice that
* Device Object (3) has resource id 13, which is "Current UNIX time of the LwM2M Client."
* Objects 3303, 3304 and 3323 have resource id 5518, which is "The timestamp of when the measurement was performed."
* Location Object (6) has resource id 5, which is "The timestamp of when the location measurement was performed."

**So it is possible to say:**

- Battery timestamp --> Object 3 resource 13
- Enviroment timestamp --> Object 3303 resource 5518 or Object 3304 resource 5518 or Object 3323 resource 5518
- GNSS timestamp --> Object 6 resource 5
- Device timestamp --> Object 3 resource 13
- Roam timestamp --> **???**

**However, [Connectivity Monitoring Object (4)](https://raw.githubusercontent.com/OpenMobileAlliance/lwm2m-registry/prod/4.xml) has not any resource related to time.**

**Comment by MLopezJ** [comment]
### Current questions 

- Is asset tracker v2 sending time values through the SenML CBOR interface? 
- > Nope. Simen said he thinks is not the case. I would confirm when I trace a send request 
- Where is the Roam timestamp data comming from?
- > Connectivity Monitoring Object (4) from LwM2M does not offer a resource to represent timestamp. It should be taken from somewhere else. This will be different as how it works in MQQT. This MUST be documented
- How one SenML CBOR request by Asset Tracker would looks like? 
- > Stig would be the one who can help me with that. Also, I can try decrypting the request with WireShark

**Comment by MLopezJ** [comment]
Previously, it was said:

> ## Uses the stablished resource from the object specification
> Some objects as Temperature (3303), Humidity (3304), Pressure (3323) and Push Button (3347), have a resource (`5518`) for inform timestamp data. There are 2 issues with this alternative:
> 
> * not all objects have this resource
> * Coiote is not reporting 5518 resource to Device Twin in azure

Now, there is known that Device Object (3) and Location Object (6) have resources related to time. And those values can be used to provided expected input in Asset Tracker web app. However, there is missing datetime value related to Roam object in the Asset Tracker web app input, and the Connectivity Monitoring Object (4) which is the one used to provide the rest of values of that object has not resources related to time. So it is still an issue. 

About the other point of Coiote not supporting timestamp related resources in Device Twin, I double checked if that is the case as well for Device Object (3) and Location Object (6) time related resources and that is not the case. 

```JavaScript

const "3" = {
                    "0": {
                        "0": {
                            "value": "Nordico"
                        },
                        "1": {
                            "value": "00010"
                        },
                        "2": {
                            "value": "00000"
                        },
                        "3": {
                            "value": "0.0"
                        },
                        "9": {
                            "value": 80
                        },
                        "13": {
                            "value": 1651820400000  // this is the resource related to time in Device
                        },
                        "16": {
                            "value": "U"
                        },
                        "18": {
                            "value": "0.0"
                        },
                        "19": {
                            "value": "0.0"
                        }
                    }
                },

const "6" = {
                    "0": {
                        "0": {
                            "value": 10.366531
                        },
                        "1": {
                            "value": -84.51215
                        },
                        "2": {
                            "value": 0
                        },
                        "3": {
                            "value": 0
                        },
                        "5": {
                            "value": 1651820400000  // this is the resource related to time in Location
                        },
                        "6": {
                            "value": 0
                        }
                    }
                },

```
However, the issue with resource id `5518` from emperature (3303), Humidity (3304), Pressure (3323) and Push Button (3347) is still there. Here https://github.com/MLopezJ/LwM2M-Asset-Tracker/issues/1 is explained why this is happening

**Comment by MLopezJ** [comment]
**It is decided to use the reported LwM2M server timestamp in case an expected timestamp is missing.**
